tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -0,0 +1,557 @@
+      hash = llvm::hash_combine(
+          hash, llvm::hash_combine_range(rhs_contracting_dims.begin(),
+                                         rhs_contracting_dims.end()));
+      hash = llvm::hash_combine(hash, mlir::hash_value(dotShape.element_type));
+      return hash;
+    }
+  };
+
+  struct BatchCluster {
+    BatchCluster(Operation* op, int op_id) : leader_op_id(op_id) {
+      ops.push_back(op);
+    }
+
+    // Merges `other` into this cluster, and clears `other`.
+    void merge(BatchCluster& other) {
+      ops.insert(ops.end(), other.ops.begin(), other.ops.end());
+      other.ops.clear();
+    }
+
+    // ID of the representative node of this cluster.
+    int leader_op_id;
+
+    // Dot ops to be batched.
+    SmallVector<Operation*> ops;
+  };
+
+  bool buildShapedDotMap();
+  bool batchingDots();
+  Value expandDim0(OpBuilder& builder, Location& loc, Value value);
+  bool applyBatching(BatchCluster& cluster);
+
+ private:
+  FuncOp func_;
+  // A map to help to cluster dot that have same shape and dim-numbers together.
+  std::unordered_map<DotShape, SmallVector<mhlo::DotGeneralOp>, DotShapeHash>
+      equal_shape_dot_map_;
+};
+
+void DotBatchingConverter::run() {
+  if (!buildShapedDotMap()) {
+    return;
+  }
+  batchingDots();
+}
+
+bool DotBatchingConverter::buildShapedDotMap() {
+  ShapeAnalysis analysis(func_);
+  if (failed(analysis.run())) {
+    LLVM_DEBUG(llvm::dbgs() << "ShapeAnalysis failes for dot merge.\n");
+    return false;
+  }
+  func_.walk([&](mhlo::DotGeneralOp op) {
+    DotShape dot_shape;
+    Value lhs = op.lhs();
+    Value rhs = op.rhs();
+    // Initialize `dimension_numbers`.
+    dot_shape.dimension_numbers = op.dot_dimension_numbers();
+    auto lhs_batch_dims =
+        dot_shape.dimension_numbers.getLhsBatchingDimensions();
+    auto rhs_batch_dims =
+        dot_shape.dimension_numbers.getRhsBatchingDimensions();
+    assert(lhs_batch_dims.size() == rhs_batch_dims.size());
+    // Initialize `batching_dims`.
+    SmallVector<DimValue>& batching_dims = dot_shape.batching_dims;
+    for (auto dim : lhs_batch_dims) {
+      batching_dims.emplace_back(std::move(analysis.getDimValue(lhs, dim)));
+    }
+    // Initialize `contracting_dim`.
+    auto lhs_contracting_dims =
+        dot_shape.dimension_numbers.getLhsContractingDimensions();
+    assert(lhs_contracting_dims.size() == 1);
+    dot_shape.contracting_dim =
+        analysis.getDimValue(lhs, lhs_contracting_dims[0]);
+    // Initialize `m_dim`.
+    int64_t lhs_rank = lhs.getType().cast<RankedTensorType>().getRank();
+    assert(lhs_batch_dims.size() + 2 == lhs_rank);
+    DenseSet<int64_t> lhs_batch_dims_set(lhs_batch_dims.begin(),
+                                         lhs_batch_dims.end());
+    for (int64_t i = 0; i < lhs_rank; i++) {
+      if ((lhs_batch_dims_set.find(i) == lhs_batch_dims_set.end()) &&
+          (i != lhs_contracting_dims[0])) {
+        dot_shape.m_dim = analysis.getDimValue(lhs, i);
```

Comment:
We only check dim-value-equality here. we maybe need to take symbolic dim equality into consideration as well.

BTW,  finding such candidate dots will became much easier  if we embed the dim-size-equality information into the tensor type.  we may re-visit this later.

```
%a = mhlo.dot_general(...) : tensor<?x?xf32, [symbolMAttr, symbolNAttr]>
...
%b = mhlo.dot_general(...) : tensor<?x?xf32, [symbolMAttr, symbolNAttr]>
```

Reply:
It is a good suggestion to use symbolic dim equality. Actually, there is no DimValue in some scenarios. As the planned reorganization for ShapeAnalysis is coming, I do not do the redundant checking. See the draft of ShapeAnalysisV2 here (which contains DimValue information within the symbolic-expr):https://github.com/JamesTheZ/BladeDISC/blob/8036255dbcee33847cb680b340675f2ce0d07703/tao_compiler/mlir/disc/transforms/shape_utils.h#L341

Reply:
> It is a good suggestion to use symbolic dim equality. Actually, there is no DimValue in some scenarios. As the planned reorganization for ShapeAnalysis is coming, I do not do the redundant checking. See the draft of ShapeAnalysisV2 here (which contains DimValue information within the symbolic-expr):https://github.com/JamesTheZ/BladeDISC/blob/8036255dbcee33847cb680b340675f2ce0d07703/tao_compiler/mlir/disc/transforms/shape_utils.h#L341

It is not hard the traceback the DimValue based symbolic-expr. Seems not necessary to introduce extra attrs.

Reply:
We can discuss this later. not blocking this PR. I personally think  type-based method is a more consistent way to represent shape constraint through different level of IR.

Reply:
> We can discuss this later. not blocking this PR. I personally think type-based method is a more consistent way to represent shape constraint through different level of IR.

Les discuss later. My point is that we may have some other approach that we even do not need to consider about the consistence.